WhateverGreen/kern_igfx.cpp

```diff
@@ -385,6 +415,75 @@ bool IGFX::wrapAcceleratorStart(IOService *that, IOService *provider) {
+	// before backlight adjustments) to grab the initial c8254 value and patch it into the controller
+	// object.
+	
+	uint64_t bxt_blc_pwm_ctrl1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_CTL1);
+	uint64_t bxt_blc_pwm_freq1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_FREQ1);
+	uint64_t bxt_blc_pwm_duty1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_DUTY1);
+	
+	DBGLOG("igfx", "wrapHwSetPanelPowerConfig(): BXT_BLC_PWM_CTRL1=0x%X BXT_BLC_PWM_FREQ1=0x%X BXT_BLC_PWM_DUTY1=0x%X", bxt_blc_pwm_ctrl1, bxt_blc_pwm_freq1, bxt_blc_pwm_duty1);
+	
+	callbackIGFX->backlightFrequency = static_cast<uint32_t>(bxt_blc_pwm_freq1);
+	
+	// This is the location of the backlight frequency. We need to store the initial value here to avoid black screen in CFL
+	uint32_t *p_fbc_pwm_freq = reinterpret_cast<uint32_t*>(reinterpret_cast<uint8_t*>(that) + 0x2b44);
+
+	if (bxt_blc_pwm_freq1)
+		*p_fbc_pwm_freq = static_cast<uint32_t>(bxt_blc_pwm_freq1);
+	
+	return FunctionCast(wrapHwSetPanelPowerConfig, callbackIGFX->orgHwSetPanelPowerConfig)(that, arg0);
+}
+
+uint64_t IGFX::wrapHwSetBacklight(void *that, uint32_t arg0) {
+	// hwSetBacklight use our backlight calculation here and skip call to original function to avoid flicker
+	//uint64_t r = FunctionCast(wrapHwSetBacklight, callbackIGFX->orgHwSetBacklight)(that, arg0);
+
+	callbackIGFX->backlightLevel = arg0;
+	
+	// uint64_t bxt_blc_pwm_ctrl1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_CTL1);
+	uint64_t bxt_blc_pwm_freq1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_FREQ1);
+	uint64_t bxt_blc_pwm_duty1 = reinterpret_cast<uint64_t(*)(void*, uint64_t)>(callbackIGFX->orgReadRegister32)(that, BXT_BLC_PWM_DUTY1);
+	
+	// Calculate backlight duty in 64-bits so result is not truncated (essentially CFL backlight fix)
+	bxt_blc_pwm_freq1 = callbackIGFX->backlightFrequency;
+	bxt_blc_pwm_duty1 = static_cast<uint64_t>(callbackIGFX->backlightLevel) * bxt_blc_pwm_freq1 / 0xFFFFLL;
+	
+	reinterpret_cast<uint64_t(*)(void*, uint64_t, uint32_t)>(callbackIGFX->orgWriteRegister32)(that, BXT_BLC_PWM_FREQ1, static_cast<uint32_t>(bxt_blc_pwm_freq1));
+	reinterpret_cast<uint64_t(*)(void*, uint64_t, uint32_t)>(callbackIGFX->orgWriteRegister32)(that, BXT_BLC_PWM_DUTY1, static_cast<uint32_t>(bxt_blc_pwm_duty1));
+	
+	// SYSLOG("igfx", "wrapHwSetBacklight(): BXT_BLC_PWM_CTRL1=0x%X BXT_BLC_PWM_FREQ1=0x%X BXT_BLC_PWM_DUTY1=0x%X backlightLevel=0x%X", bxt_blc_pwm_ctrl1, bxt_blc_pwm_freq1, bxt_blc_pwm_duty1, callbackIGFX->backlightLevel);
+	
+	return 0LL;
+}
+
+uint64_t IGFX::wrapSetAttributeForConnection(void *that, uint32_t arg0, uint32_t arg1, uint64_t arg2) {
+	// setAttributeForConnection appears to be called from several places and will cause black screen bug in CFL if
```

Comment:
Same thing about `uint64_t`, and please make it more clear, that the real issue is 32-bit integer truncation in Apple code due to multiplying by a value being too large than usual on CFL. I do not remember the exact values, but was it `0xFFFFF`? Probably worth including sample values here too.